core/src/androidMain/kotlin/Connection.kt

```diff
@@ -0,0 +1,71 @@
+package com.juul.kable
+
+import android.bluetooth.BluetoothGatt
+import android.bluetooth.BluetoothGatt.GATT_SUCCESS
+import android.os.RemoteException
+import com.juul.kable.gatt.Callback
+import com.juul.kable.gatt.GattStatus
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+
+public class OutOfOrderGattCallbackException internal constructor(
+    message: String,
+) : IllegalStateException(message)
+
+internal data class CharacteristicChange(
+    val characteristic: Characteristic,
+    val data: ByteArray,
+)
+
+private val GattSuccess = GattStatus(GATT_SUCCESS)
+
+internal class Connection(
+    internal val bluetoothGatt: BluetoothGatt,
+    private val dispatcher: CoroutineDispatcher,
+    private val callback: Callback,
+    private val invokeOnClose: () -> Unit,
+) {
+
+    init {
+        callback.invokeOnDisconnected(::close)
+    }
+
+    val characteristicChanges = callback.onCharacteristicChanged
+        .map { (bluetoothGattCharacteristic, value) ->
+            CharacteristicChange(bluetoothGattCharacteristic.toLazyCharacteristic(), value)
+        }
+
+    private val lock = Mutex()
+
+    suspend inline fun <reified T> request(
+        crossinline action: BluetoothGatt.() -> Boolean,
```

Comment:
Know this is only used internally but some documentation around 'action' seems warranted

Reply:
Agreed. I'll add some documentation. 👍